test/phabricator_test/test_phabricator.py

```diff
@@ -0,0 +1,356 @@
+        response = PhabricatorService().time_from_epoch(123456789)
+        self.assertEqual(response, expected_response)
+
+    def test_differential_query_successful(self):
+        """
+            Tests differential query API Call (successful)
+        """
+        expected_response = "Successful call!"
+
+        fake_phab = MagicMock()
+        fake_phab.differential.query.return_value = "Successful call!"
+        fake_status = "Open"
+        fake_responsibleUsers = ["PHID-1", "PHID-2"]
+
+        response = PhabricatorService().differential_query(
+            status=fake_status,
+            responsibleUsers=fake_responsibleUsers,
+            phab=fake_phab
+        )
+        fake_phab.differential.query.assert_called_with(
+            status=fake_status,
+            responsibleUsers=fake_responsibleUsers
+        )
+        self.assertEqual(expected_response, response)
+
+    # TODO: Very similar methods (user_query_ids, user_query_usernames etc. That will be based off 'test_differential_query'.
+
+    @patch(PATH + 'PhabricatorService.user_query_ids')
+    def test_author_data_in_rawReponse(self, mock_user_query_ids):
+        """
+            Tests author_data function where author_phid IS IN the raw response
+        """
+
+        fake_user, fake_raw_response = self.mock.mock_phabricator_author_data_rawResponse()
+
+        user, raw_response = PhabricatorService().author_data(
+            author_phid=fake_user['phid'],
+            raw_response=fake_raw_response,
+            phab=None
+        )
+
+        mock_user_query_ids.assert_not_called()
+        self.assertEqual(fake_user, user)
+        self.assertEqual(fake_raw_response, raw_response)
+
+    @patch(PATH + 'PhabricatorService.user_query_ids',
+           return_value=["New User!"])
+    def test_author_data_not_in_rawResponse(self, mock_user_query_ids):
+        """
+            Tests author_data function where author_phid IS NOT IN the raw response
+        """
+        expected_raw_response = ["New User!"]
+        expected_user = "New User!"
+
+        fake_user, _ = self.mock.mock_phabricator_author_data_rawResponse()
+        
+        user, raw_response = PhabricatorService().author_data(
+            author_phid=fake_user['phid'],
+            raw_response=[],
+            phab=None
+        )
+
+        mock_user_query_ids.assert_called_with(
+            [fake_user['phid']],
+            None
+        )
+        self.assertEqual(expected_user, user)
+        self.assertEqual(expected_raw_response, raw_response)
+
+    @patch(PATH + 'PhabricatorService.user_query_ids',
+           side_effect=Exception('404 Error Exception'))
+    def test_author_data_failure(self, mock_user_query_ids):
+        """
+            Tests author_data function where user_query_ids throws an error (i.e. 404)
+        """
+
+        fake_user, _ = self.mock.mock_phabricator_author_data_rawResponse()
+
+        with self.assertRaises(Exception) as context:
```

Comment:
There is a catch. If something else screwed up, for example KeyError was raised, this will pass even if this test should pass only if Exception('404 Error Exception') is raised.

Reply:
I don't think this test is actually required since you dont manipulate with the exception. This will do as much good as putting wrong dictionary as `raw_response` and verifying that KeyError is raised.

Reply:
I will remove this test but still had a question. As context is the catch, under that function I have: 

` self.assertIn('404 Error Exception', context.exception)`

Wouldn't this ensure that the test will only pass i the '404 Error Exception' is raised?

Reply:
You could do that like this, however much better option would be to create a custom exception
```python
class MyCustomException(Exception):
    pass
```
then set it as side effect and finally check that raised exception is really the one set as side effect.
```python
with self.assertRaises(MyCustomException):
    ...
```
I like this approach more because you don't need to look at the text in the exception.

Reply:
Feel free to mark this conversation as resolved.